storagemarket/impl/providerstates/provider_states.go

```diff
@@ -266,13 +267,50 @@ func HandoffDeal(ctx fsm.Context, environment ProviderDealEnvironment, deal stor
 		paddedSize,
 		paddedReader,
 	)
-
 	if err != nil {
 		return ctx.Trigger(storagemarket.ProviderEventDealHandoffFailed, err)
 	}
+
+	if err := registerDealDataForEarlyRetrieval(environment, deal, packingInfo.SectorNumber, packingInfo.Offset, packingInfo.Size); err != nil {
+		log.Errorf("failed to register deal data for early retrieval: %s", err)
+	}
+
 	return ctx.Trigger(storagemarket.ProviderEventDealHandedOff)
 }
 
+func registerDealDataForEarlyRetrieval(environment ProviderDealEnvironment, deal storagemarket.MinerDeal, sectorID abi.SectorNumber, offset, length abi.PaddedPieceSize) error {
```

Comment:
is this the same code as in RecordPieceInfo -- we should reuse it then no?

Also, it looks like there is no check on whether it's been called already there -- is that going to cause problems?